sia/src/rhp/merkle.rs

```diff
@@ -73,6 +75,149 @@ pub fn sector_root(sector: &[u8]) -> Hash256 {
     )
 }
 
+#[derive(Debug, Error)]
+pub enum ProofValidationError {
+    #[error("Invalid proof length: expected {expected}, got {actual}")]
+    InvalidProofLength { expected: usize, actual: usize },
+
+    #[error("Invalid proof root: expected {expected}, got {actual}")]
+    InvalidProofRoot { expected: Hash256, actual: Hash256 },
+
+    #[error("IO error: {0}")]
+    Io(#[from] io::Error),
+
+    #[error("The proof data is not segment aligned")]
+    NotSegmentAligned,
+}
+
+pub type Result<T> = std::result::Result<T, ProofValidationError>;
+
+pub struct RangeProof;
+
+impl RangeProof {
+    #[allow(dead_code)]
+    pub async fn verify<R: AsyncRead + Unpin>(
+        r: R,
+        root: &Hash256,
+        proof: &[Hash256],
+        start: usize,
+        end: usize,
+    ) -> Result<()> {
+        let roots = Self::read_data(&mut io::BufReader::new(r), start, end).await?;
+
+        if proof.len() != range_proof_size(LEAVES_PER_SECTOR, start, end) {
+            return Err(ProofValidationError::InvalidProofLength {
+                expected: range_proof_size(LEAVES_PER_SECTOR, start, end),
+                actual: proof.len(),
+            });
+        }
+
+        let consume = |acc: &mut Accumulator, roots: &[Hash256], i: usize, j: usize| {
+            let mut roots = roots;
+            let mut i = i;
+            while i < j && !roots.is_empty() {
+                let subtree_size = next_subtree_size(i, j);
+                let height = subtree_size
+                    .checked_ilog2()
```

Comment:
Not sure what's better. This seems more correct, but `trailing_zeros` doesn't return an option 🤷🏻